pkg/cli/alpha/config-gen/types.go

```diff
@@ -0,0 +1,192 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package configgen
+
+import (
+	"io/ioutil"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/kustomize/kyaml/errors"
+)
+
+// KubebuilderConfigGen implements the API for generating configuration
+type KubebuilderConfigGen struct {
+	metav1.TypeMeta   `json:",inline" yaml:",omitempty"`
+	metav1.ObjectMeta `json:"metadata,omitempty" yaml:"metadata,omitempty"`
+
+	// Spec is the configuration spec defining what configuration should be produced.
+	Spec KubebuilderConfigGenSpec `json:"spec,omitempty" yaml:"spec,omitempty"`
+
+	// Status is the configuration status defined at runtime.
+	Status KubebuilderConfigGenStatus `json:"status,omitempty" yaml:"status,omitempty"`
+
+	// Project is for the field name used by the PROJECT file (config), instead of spec.
+	// Use Spec unless in the PROJECT file.
+	Project *KubebuilderConfigGenSpec `json:"config,omitempty" yaml:"config,omitempty"`
+}
+
+var (
+	// NamespaceComponent will create the namespace for the controller-manager
+	NamespaceComponent = "namespace"
+
+	// ControllerManagerComponent will create the controller-manager Deployment
+	ControllerManagerComponent = "controller-manager"
+
+	// WebhooksComponent will create the webhook configurations
+	WebhooksComponent = "webhooks"
+
+	// CRDsComponent will create the CRDs
+	CRDsComponent = "crds"
+
+	// RBACComponent will create RBAC rules
+	RBACComponent = "rbac"
+
+	// CertManagerComponent will create the Issuer and Certificate resources
+	// for CertManager to inject the certificates
+	CertManagerComponent = "cert-manager"
+
+	// PrometheusComponent will create the ServiceMonitor resource
+	PrometheusComponent = "prometheus"
+)
+
+// getDefaultComponents returns the set of components that are created by default
+func getDefaultComponents() map[string]bool {
+	return map[string]bool{
+		NamespaceComponent:         true,
+		ControllerManagerComponent: true,
+		RBACComponent:              true,
+		CRDsComponent:              true,
+	}
+}
+
+// Enabled returns true if the component is enabled
+func (a *KubebuilderConfigGenSpec) Enabled(component string) bool {
+	return a.Components[component]
+}
+
+// ComponentConfigEnabled returns true if component config is being used
+func (a *KubebuilderConfigGenSpec) ComponentConfigEnabled() bool {
+	return a.ComponentConfigFilepath != ""
+}
```

Comment:
Does it make sense conceptually to have `ComponentConfig` special-cased? Maybe we can have two maps (one for booleans and other for strings) to enable other future components that store a string and use a single method that checks for `true` in one or for `!= ""` in the other.

Reply:
P.S.: Am I correct thinking that this is a map instead of an struct to allow programatic addition to fields here? Like for example extensions defining new components?

Reply:
Is see this as different.  Enabling component config requires setting the file path rather than setting it simply as on.

The Components could be a struct, I'm not sure if it matters too much.  It is a collection of key-value pairs for flagging individual components as enabled or disabled.

Reply:
The point I wanted to raise was, may there be additional components that require a path instead of an "on" status? And, does it make sense to have something like the following snippet?
```go
// Enabled returns true if the component is enabled
func (a *KubebuilderConfigGenSpec) Enabled(component string) bool {
	if enabled, found := a.Components[component]; found {
		return enabled
	}
	if path, found := a.ComponentsWithPath[component]; found {
		return path != ""
	}
	return false
}
```

Reply:
It just feels a bit odd to have components enabled and disabled in different levels:
```yaml
components:
  namespace: true
  componentConfig: true
disableAuthProxy: true
componentConfigFilepath: "..."
```
and then checked with different functions.

Using a struct would make the `Enabled` function more verbose and would hard-code the available components but if those are non-issues it could be done something like:
```go
type Components struct {
	Namespace           bool   `json:"enableNamespace,omitempty"`         `yaml:"enableNamespace,omitempty"`
	ControllerManager   bool   `json:"enableControllerManager,omitempty"` `yaml:"enableControllerManager,omitempty"`
	NoAuthProxy         bool   `json:"disableAuthProxy,omitempty"`        `yaml:"disableAuthProxy,omitempty"`
	ComponentConfigPath string `json:"pathForComponentConfig,omitempty"`  `yaml:"pathForComponentConfig,omitempty"`
}

func (c Componentes) Enabled(component string) bool {
	switch component {
	case NamespaceComponent:
		return c.Namespace
	case ControllerManagerComponent:
		return c.ControllerManager
	case AuthProxyComponent:
		return !c.NoAuthProxy
	case ComponentConfigComponent:
		return c.ComponentConfigPath != ""
	default:
		return false
	}
}
```

and the resulting yaml
```yaml
components:
  enableNamespace: true
  enableComponentConfig: true
  disableAuthProxy: true
  pathForComponentConfig: "..."
```

P.S.: I'm using an slightly different nomenclature to explicitly add path and enabled to the fields but feel free to use whatever nomenclature you feel first better.

Reply:
I am happy to change the API however we want.

As it stands today, this API makes a distinction between a component (*what*) and feature (*how*).  Features exist to support components and don't make sense on their own.

It isn't a hard distinction -- would just generating the a namespace make sense?  IDK.  But the namespace more or less stands alone and is loosely coupled enough to the other components that I think it could be.  However just generating authproxy doesn't make sense to me -- authproxy is defined by rbac roles, services, and patches to other resources.  The same is true for componentConfig.

We could remove the concept of components altogether and just make things features.  The intent was to break the config into loosely coupled groups of resource that could be generated independently of one another, but that may not make sense.  Is there a use case for *not* creating the namespace or controller-manager.  Probably not worth implementing for alpha in any case.

Why don't I simply drop the components field and always generate all the resources?